js.go

```diff
@@ -1214,6 +1364,262 @@ type StreamState struct {
+	if err != nil {
+		return nil, err
+	}
+
+	ssSubj := js.apiSubj(fmt.Sprintf(JSApiStreamSnapshotT, name))
+	r, err := js.nc.Request(ssSubj, req, js.wait)
+	if err != nil {
+		return nil, err
+	}
+	var resp JSApiStreamSnapshotResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return nil, err
+	}
+	if resp.Error != nil {
+		return nil, errors.New(resp.Error.Description)
+	}
+
+	// Setup to process snapshot chunks.
+	var snapshot []byte
+	done := make(chan bool)
+	sub, err := js.nc.Subscribe(cfg.DeliverSubject, func(m *Msg) {
+		// EOF
+		if len(m.Data) == 0 {
+			done <- true
+			return
+		}
+		// Could be writing to a file here too.
+		snapshot = append(snapshot, m.Data...)
+		// Flow ack
+		m.Respond(nil)
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer sub.Unsubscribe()
+	// Wait to receive the snapshot.
+	select {
+	case <-done:
+	case <-time.After(5 * time.Second):
+		return nil, errors.New("nats: snapshot timeout exceeded")
+	}
+
+	return bytes.NewReader(snapshot), nil
+}
+
+// JSApiStreamRestoreResponse is the direct response to the restore request.
+type JSApiStreamRestoreResponse struct {
+	APIResponse
+	// Subject to deliver the chunks to for the snapshot restore.
+	DeliverSubject string `json:"deliver_subject"`
+}
+
+func (js *js) RestoreStream(name string, snapshot io.Reader) error {
+	rsSubj := js.apiSubj(fmt.Sprintf(JSApiStreamRestoreT, name))
+	r, err := js.nc.Request(rsSubj, nil, js.wait)
+	if err != nil {
+		return err
+	}
+	var resp JSApiStreamRestoreResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return err
+	}
+	if resp.Error != nil {
+		return errors.New(resp.Error.Description)
+	}
+
+	// Can be any size message.
+	var chunk [512]byte
+	for r := snapshot; ; {
+		n, err := r.Read(chunk[:])
+		if err != nil {
+			break
+		}
+		js.nc.Request(resp.DeliverSubject, chunk[:n], time.Second)
+	}
+	js.nc.Request(resp.DeliverSubject, nil, time.Second)
+	return nil
+}
+
+// JSApiStreamPurgeResponse.
+type JSApiStreamPurgeResponse struct {
+	APIResponse
+	Success bool   `json:"success,omitempty"`
+	Purged  uint64 `json:"purged"`
+}
+
+func (js *js) PurgeStream(name string) error {
+	psSubj := js.apiSubj(fmt.Sprintf(JSApiStreamPurgeT, name))
+	r, err := js.nc.Request(psSubj, nil, js.wait)
+	if err != nil {
+		return err
+	}
+	var resp JSApiStreamPurgeResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return err
+	}
+	if resp.Error != nil {
+		return errors.New(resp.Error.Description)
+	}
+	return nil
+}
+
+type JSApiStreamNamesRequest struct {
+	APIPagedRequest
+	// These are filters that can be applied to the list.
+	Subject string `json:"subject,omitempty"`
+}
+
+// JSApiStreamListResponse list of detailed stream information.
+// A nil request is valid and means all streams.
+type JSApiStreamListResponse struct {
+	APIResponse
+	APIPaged
+	Streams []*StreamInfo `json:"streams"`
+}
+
+// StreamLister fetches pages of StreamInfo objects.
+type StreamLister struct {
```

Comment:
Same comment than for the consumer lister. Maybe add a note that this is not thread safe?